packages/react-reconciler/src/ReactMutableSource.js

```diff
@@ -30,20 +30,42 @@ export function clearPendingUpdates(
   root: FiberRoot,
   expirationTime: ExpirationTime,
 ): void {
-  if (root.mutableSourcePendingUpdateTime <= expirationTime) {
-    root.mutableSourcePendingUpdateTime = NoWork;
+  if (root.mutableSourceFirstPendingUpdateTime <= expirationTime) {
+    root.mutableSourceFirstPendingUpdateTime = NoWork;
   }
+  if (root.mutableSourceLastPendingUpdateTime <= expirationTime) {
+    root.mutableSourceLastPendingUpdateTime = NoWork;
+  }
+}
+
+export function getFirstPendingExpirationTime(root: FiberRoot): ExpirationTime {
+  return root.mutableSourceFirstPendingUpdateTime;
 }
 
-export function getPendingExpirationTime(root: FiberRoot): ExpirationTime {
-  return root.mutableSourcePendingUpdateTime;
+export function getLastPendingExpirationTime(root: FiberRoot): ExpirationTime {
+  return root.mutableSourceLastPendingUpdateTime;
 }
 
 export function setPendingExpirationTime(
   root: FiberRoot,
   expirationTime: ExpirationTime,
 ): void {
-  root.mutableSourcePendingUpdateTime = expirationTime;
+  // Because we only track one pending update time per root,
+  // track the lowest priority update.
+  // It's inclusive of all other pending updates.
+  root.mutableSourceLastPendingUpdateTime = Math.max(
+    root.mutableSourceLastPendingUpdateTime,
+    expirationTime,
+  );
+
+  if (root.mutableSourceFirstPendingUpdateTime === NoWork) {
+    root.mutableSourceFirstPendingUpdateTime = expirationTime;
+  } else {
+    root.mutableSourceFirstPendingUpdateTime = Math.min(
```

Comment:
Nit: use comparison operators instead of Math.max like we do elsewhere. Easier to optimize. As an example, note how this does an assignment every time this function is called, even if the value isn't changed.